packages/core/src/engine/Optimizer.ts

```diff
@@ -815,7 +816,14 @@ export const genOptProblem = (
       gradf: xs.map((x, i) => {
         // fill in any holes in case some inputs weren't used in the graph, and
         // also treat pending values as constants rather than optimizing them
-        return i in gradient && !("pending" in inputs[i]) ? gradient[i] : 0;
+        if (!(i in gradient)) {
+          return 0;
+        } else {
+          const meta = inputs[i];
+          return meta.tag === "Optimized" && !frozenVaryingValues.includes(i)
```

Comment:
Actually, given that we use `includes` here, why not make `frozenVaryingValues` a [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead of an array?